pyro/infer/mcmc/hmc.py

```diff
@@ -210,7 +214,12 @@ def sample(self, trace):
             z = z_new
 
         if self.adapt_step_size:
-            accept_prob = (-delta_energy).exp().clamp(max=1).item()
+            # Set accept prob to 0.0 if delta_energy is `NaN` which could be the case
+            # which may be the case for a diverging trajectory (e.g. in the case of
+            # evaluating log prob of a value simulated using a large step size for
+            # a constrained sample site).
+            accept_prob = (-delta_energy).exp().clamp(max=1).item() if not torch_isnan(delta_energy) \
+                else torch.tensor(0.0)
```

Comment:
This will not work with GPUs. Instead use `.new_zeros()` or `.new_tensor()`
```py
if torch_isnan(...):
    accept_prob = delta_energy.new_tensor(0.0)
else:
    accept_prob  = ...
```